dotnet/src/Agents/OpenAI/Internal/AssistantMessageFactory.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+using System.Collections.Generic;
+using OpenAI.Assistants;
+
+namespace Microsoft.SemanticKernel.Agents.OpenAI.Internal;
+
+/// <summary>
+/// Factory for creating <see cref="MessageContent"/> based on <see cref="ChatMessageContent"/>.
+/// Also able to produce <see cref="MessageCreationOptions"/>.
+/// </summary>
+/// <remarks>
+/// Improves testability.
+/// </remarks>
+internal static class AssistantMessageFactory
+{
+    /// <summary>
+    /// Produces <see cref="MessageCreationOptions"/> based on <see cref="ChatMessageContent"/>.
+    /// </summary>
+    /// <param name="message">The message content.</param>
+    public static MessageCreationOptions CreateOptions(ChatMessageContent message)
+    {
+        MessageCreationOptions options = new();
+
+        if (message.Metadata != null)
+        {
+            foreach (var metadata in message.Metadata)
+            {
+                options.Metadata.Add(metadata.Key, metadata.Value?.ToString() ?? string.Empty);
+            }
+        }
+
+        return options;
+    }
+
+    /// <summary>
+    /// Translates <see cref="ChatMessageContent.Items"/> into enumeration of <see cref="MessageContent"/>.
+    /// </summary>
+    /// <param name="message">The message content.</param>
+    public static IEnumerable<MessageContent> GetMessageContents(ChatMessageContent message)
+    {
+        foreach (KernelContent content in message.Items)
+        {
+            if (content is TextContent textContent)
+            {
+                yield return MessageContent.FromText(content.ToString());
+            }
+            else if (content is ImageContent imageContent)
+            {
+                if (imageContent.Uri != null)
+                {
+                    yield return MessageContent.FromImageUrl(imageContent.Uri);
+                }
+                //else if (string.IsNullOrWhiteSpace(imageContent.DataUri))
+                //{
+                //    SDK BUG - BAD SIGNATURE (https://github.com/openai/openai-dotnet/issues/135)
+                //        URI does not accept the format used for `DataUri`
+                //        Approach is inefficient anyway...
+                //    yield return MessageContent.FromImageUrl(new Uri(imageContent.DataUri!));
+                //}
```

Comment:
If you read the bug, we'll never use `FromImageUrl` to accomodate this pattern.  The plan is a new factory method will be added to `MessageContent` that takes a `BinaryData`.  The core issue here is that `Uri` constructor will _never_ accept the `DataUri` format (so the exception thrown will be quite misleading).

In the spirit of your comment, we can throw `KernelException` instead of silent failure.

Reply:
Updated

Reply:
Ok.  I could've sworn I got a constructor exception on the scheme...but I can see your are correct.  Updated.